File: ruby/6kyu/are_they_the_same.rb
# Given two arrays a and b write a function comp(a, b) (compSame(a, b) in Clojure) that checks whether the two arrays have the 
# "same" elements, with the same multiplicities. "Same" means, here, that the elements in b are the elements in a squared, 
# regardless of the order.

# initial attempt:
# def comp(array1, array2)
#   if (array1 == nil) || (array2 == nil) 
#     false
#   elsif (array1.empty?) || (array2.empty?)
#     "a or b are empty or not empty lists."
#   else
#     array2.map {|number| 
#       number / number
#     }
#     array1.sort == array2.sort
#   end
# end

# 2nd attempt
def comp(a, b)
  (a == nil) || (b == nil) ? false : a.sort.map { |num| num ** 2} == b.sort
end

p comp([121, 144, 19, 161, 19, 144, 19, 11], [132, 14641, 20736, 361, 25921, 361, 20736, 361])

File: javascript/6kyu/areTheyTheSame.js
/*
Given two arrays a and b write a function comp(a, b)
(orcompSame(a, b)) that checks whether the two
arrays have the "same" elements, with the same
multiplicities. "Same" means, here, that the elements
in b are the elements in a squared, regardless of
the order.
*/

const squareElements = (array) => array.map(number  => number * number);
const leastToGreatest = (a, b) => a - b;

function comp(array1, array2){
  if (!array1 || !array2) return false;
  
  const elementsSquared = squareElements(array1).sort(leastToGreatest);
  const array2Sorted = array2.sort(leastToGreatest);

  return elementsSquared.every((number, index) => number === array2Sorted[index]);
}

console.log(comp([121, 144, 19, 161, 19, 144, 19, 11], [11*11, 121*121, 144*144, 19*19, 161*161, 19*19, 144*144, 19*19]));
// true

console.log(comp(
  [ 121, 144, 19, 161, 19, 144, 19, 11 ],
  [ 121, 14641, 20736, 36100, 25921, 361, 20736, 361 ]));
// false
